packages/serializer/source/serializer.ts

```diff
@@ -37,32 +38,69 @@ export class Serializer implements Contracts.Serializer.ISerializer {
 				continue;
 			}
 
+			const value = data[property];
+			if (schema.required) {
+				Utils.assert.defined(value);
+			}
+
 			if (schema.type === "uint32") {
-				result.writeUint32(data[property]);
+				result.writeUint32(value);
+				continue;
 			}
 
 			if (schema.type === "uint64") {
-				result.writeUint64(data[property]);
+				result.writeUint64(value);
+				continue;
 			}
 
 			if (schema.type === "bigint") {
-				result.writeUint64(data[property].toString());
+				result.writeUint64(value.toString());
+				continue;
 			}
 
 			if (schema.type === "hash") {
-				result.writeBytes(Buffer.from(data[property], "hex"));
+				result.writeBytes(Buffer.from(value, "hex"));
+				continue;
+			}
+
+			if (schema.type === "blockId") {
+				if (value === undefined) {
+					const nullBlockId = "0000000000000000000000000000000000000000000000000000000000000000";
+					result.writeBytes(Buffer.from(nullBlockId, "hex"));
+				} else {
+					result.writeBytes(Buffer.from(value, "hex"));
+				}
+
+				continue;
 			}
 
 			if (schema.type === "address") {
-				this.addressSerializer.serialize(result, await this.addressFactory.toBuffer(data[property]));
+				this.addressSerializer.serialize(result, await this.addressFactory.toBuffer(value));
+				continue;
 			}
 
 			if (schema.type === "publicKey") {
 				this.publicKeySerializer.serialize(result, data[property]);
+				continue;
 			}
 
 			if (schema.type === "signature") {
-				this.signatureSerializer.serialize(result, data[property]);
+				if (value === undefined) {
```

Comment:
Signature cannot be undefined. This is different use case. 

```
...(options.excludeSignature
? {}
: {
	signature: {
		type: "signature",
	},
	}),
```

This code represent if the signature is part of schema (and not the part of serialzier options). If it is present is always defined.